machine: refactor pins to be of Pin type

aykevl opened the pull request:
Here is an API-breaking refactor I've long wanted to do. Basically, instead of having GPIO objects, all pins are represented by a Pin object (usually an `int8`, might be different on boards with _lots_ of pins). This has several advantages:

 * Pin numbers can have the type `Pin`, which immediately gives them the common `Configure`, `Set`, etc. methods. No need for `GPIO{rx}.Configure` etc.
 * I think Pin just looks nicer than GPIO. MicroPython has the same Pin type (as `machine.Pin`, actually).

Comments welcome. This PR is not nearly complete: the things that should work are some nrf-based boards and the blinky1/blinky2/echo examples.

aykevl commented:
Not entirely finished (needs testing) but ready for review.

aykevl commented:
Also, if we agree to merge this, should this happen before or after the next release?

deadprogram commented on src/machine/machine_atsamd21.go:
```diff
@@ -18,113 +18,93 @@ import (
-	GPIO_PWM_ALT        = GPIO_TIMER_ALT
-	GPIO_INPUT_PULLDOWN = 12
+	PinAnalog    PinMode = 1
+	PinSerCom    PinMode = 2
```
Probably should name this `PinSERCOM` only because the datasheet and all other all use all upper case.

aykevl replied:
Yes, this was one of the things I was not so sure about.

deadprogram replied:
Either that or `PinSercom`. `PinSerCom` looks very odd.

aykevl replied:
SERCOM is an abbreviation of Serial Communication, so that's why I opted for SerCom.

Anyway, SERCOM is probably better.

deadprogram commented on src/machine/machine_atsamd21.go:
```diff
@@ -18,113 +18,93 @@ import (
-	GPIO_INPUT_PULLDOWN = 12
+	PinAnalog    PinMode = 1
+	PinSerCom    PinMode = 2
+	PinSerComAlt PinMode = 3
```
Probably should name this `PinSERCOMAlt` only because the datasheet and all other all use all upper case for the `SERCOM`.

deadprogram commented:
This is a breaking change, so there is no good time. In other words, I think we are better off doing it now rather than later. 

A number of the drivers will require changes at the same time, but the nature of the change is not large, so it can be done pretty quickly.

aykevl commented on src/machine/machine_nrf.go:
```diff
@@ -7,25 +7,25 @@ import (
+func (p Pin) Configure(config PinConfig) {
 	cfg := config.Mode | nrf.GPIO_PIN_CNF_DRIVE_S0S1 | nrf.GPIO_PIN_CNF_SENSE_Disabled
 	port, pin := p.getPortPin()
 	port.PIN_CNF[pin].Set(uint32(cfg))
```
Interesting, this function gets inlined after this change resulting in reduced code size, because this whole function including `getPortPin` is reduced to a single instruction in the caller. Probably because `getPortPin` avoids an extractvalue (which already was a no-op basically) which tipped the balance towards having this one inlined.

aykevl commented:
I have compared all programs compiled in `make smoketest` and they all are the same size or slightly reduced in size. In fact, the majority was reduced in size. That's surprising to me, as I did not expect any code size changes, or any significant code changes at all.

deadprogram commented:
Interesting. I suppose some of the code is recognized by the optimizer as being the same?

Have you tested anything besides the basic GPIO with this code yet?

aykevl commented:
I made some changes, but verified that they did not change any of the smoke tests (by comparing the checksums of output hex files).

* renamed `SerCom` to `SERCOM`.
* added `machine.NoPin` (similar to [token.NoPos](https://golang.org/pkg/go/token/#Pos)), to make it more obvious that some pins are not set. Using `machine.NoPin` also makes it easier to change the value of this (currently -1) if that is ever required.
* added a few documentation comments.

aykevl commented:
I haven't really tested anything on real hardware yet, will need to do that.

deadprogram commented:
Just tested on ItsyBitsy-M0 with blinky1 and also GPS example from drivers repo, and both worked perfectly.

deadprogram commented:
Tested a few other examples, and also now have a PR ready on the drivers repo to match these changes. I say we do it!

aykevl commented:
Ok, so ready for a merge now? Or is there anything we need to test before that?
I can do some testing but probably have less hardware close by.

deadprogram commented:
I am going to merge now, then after the docker build has a chance to finish, I will merge the associated PR in the drivers directory.

We should also update the wiki to match https://github.com/tinygo-org/tinygo/wiki/Machine-API#pins

aykevl commented:
Great!
Updated the wiki.
